lib/vnet/setup_darwin.go

```diff
@@ -0,0 +1,254 @@
+		return nil, "", trace.Wrap(err, "setting permissions on unix socket")
+	}
+	return l, socketPath, nil
+}
+
+// sendTUNNameAndFd sends the name of the TUN device and its open file descriptor over a unix socket, meant
+// for passing the TUN from the root process which must create it to the user process.
+func sendTUNNameAndFd(socketPath, tunName string, fd uintptr) error {
+	socketAddr := &net.UnixAddr{Name: socketPath, Net: "unix"}
+	conn, err := net.DialUnix(socketAddr.Net, nil /*laddr*/, socketAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	err = conn.SetDeadline(time.Now().Add(tunHandoverTimeout))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Write the device name as the main message and pass the file desciptor as out-of-band data.
+	rights := unix.UnixRights(int(fd))
+	if _, _, err := conn.WriteMsgUnix([]byte(tunName), rights, socketAddr); err != nil {
+		return trace.Wrap(err, "writing to unix conn")
+	}
+	return nil
+}
+
+// sendTUNNameAndFd receives the name of a TUN device and its open file descriptor over a unix socket, meant
+// for passing the TUN from the root process which must create it to the user process.
+func recvTUNNameAndFd(ctx context.Context, socket *net.UnixListener) (string, uintptr, error) {
+	ctx, cancel := context.WithTimeout(ctx, tunHandoverTimeout)
+	defer cancel()
+	deadline, _ := ctx.Deadline()
+
+	err := socket.SetDeadline(deadline)
+	if err != nil {
+		return "", 0, trace.Wrap(err)
+	}
+	go func() {
+		<-ctx.Done()
+		socket.Close()
+	}()
+
+	conn, err := socket.AcceptUnix()
+	if err != nil {
+		return "", 0, trace.Wrap(err)
+	}
+	go func() {
+		// Close the connection early to unblock reads if the context is canceled.
+		<-ctx.Done()
+		conn.Close()
+	}()
+
+	msg := make([]byte, 128)
+	oob := make([]byte, unix.CmsgSpace(4)) // Fd is 4 bytes
+	n, oobn, _, _, err := conn.ReadMsgUnix(msg, oob)
+	if err != nil {
+		return "", 0, trace.Wrap(err, "reading from unix conn")
+	}
+
+	// Parse the device name from the main message.
+	if n == 0 {
+		return "", 0, trace.Errorf("failed to read msg from unix conn")
+	}
+	if oobn != len(oob) {
+		return "", 0, trace.Errorf("failed to read out-of-band data from unix conn")
+	}
+	tunName := string(msg[:n])
+
+	// Parse the file descriptor from the out-of-band data.
+	scm, err := unix.ParseSocketControlMessage(oob)
+	if err != nil {
+		return "", 0, trace.Wrap(err, "parsing socket control message")
+	}
+	if len(scm) != 1 {
+		return "", 0, trace.BadParameter("expect 1 socket control message, got %d", len(scm))
+	}
+	fds, err := unix.ParseUnixRights(&scm[0])
+	if err != nil {
+		return "", 0, trace.Wrap(err, "parsing file descriptors")
+	}
+	if len(fds) != 1 {
+		return "", 0, trace.BadParameter("expected 1 file descriptor, got %d", len(fds))
+	}
+	fd := uintptr(fds[0])
+
+	return tunName, fd, nil
+}
+
+func configureOS(ctx context.Context, cfg *osConfig) error {
+	if cfg.tunIPv6 != "" && cfg.tunName != "" {
```

Comment:
Consider inverting the condition here so that the `return nil` is indented and the body that does most of the work is not in a conditional.

Reply:
I'm going to be adding more conditions later on - see https://github.com/gravitational/teleport/pull/41032/files#diff-7627839bca33fec5fb081fb7297edbdbaf62435ee4e8d8105c76ccbb7edf8150